library/src/nvcc_detail/hipsparse.cpp

```diff
@@ -9341,6 +9341,60 @@ cusparseSpMVAlg_t hipSpMVAlgToCudaSpMVAlg(hipsparseSpMVAlg_t alg)
 }
 #endif
 
+#if(CUDART_VERSION >= 11000)
+cusparseSpMMAlg_t hipSpMMAlgToCudaSpMMAlg(hipsparseSpMMAlg_t alg)
+{
+    switch(alg)
+    {
+    case HIPSPARSE_MM_ALG_DEFAULT:
+        return CUSPARSE_MM_ALG_DEFAULT;
+    case HIPSPARSE_COOMM_ALG1:
+        return CUSPARSE_COOMM_ALG1;
+    case HIPSPARSE_COOMM_ALG2:
+        return CUSPARSE_COOMM_ALG2;
+    case HIPSPARSE_COOMM_ALG3:
+        return CUSPARSE_COOM_ALG3;
+    case HIPSPARSE_CSRMM_ALG1:
+        return CUSPARSE_CSRMM_ALG1;
+    case HIPSPARSE_SPMM_ALG_DEFAULT:
+        return CUSPARSE_SPMM_ALG_DEFAULT;
+    case HIPSPARSE_SPMM_COO_ALG1:
+        return CUSPARSE_SPMM_COO_ALG1;
+    case HIPSPARSE_SPMM_COO_ALG2:
+        return CUSPARSE_SPMM_COO_ALG2;
+    case HIPSPARSE_SPMM_COO_ALG3:
+        return CUSPARSE_SPMM_COO_ALG3;
+    case HIPSPARSE_SPMM_COO_ALG4:
+        return CUSPARSE_SPMM_COO_ALG4;
+    case HIPSPARSE_SPMM_CSR_ALG1:
+        return CUSPARSE_SPMM_CSR_ALG1;
+    case HIPSPARSE_SPMM_CSR_ALG2:
+        return CUSPARSE_SPMM_CSR_ALG2;
+    default:
```

Comment:
I would remove the default case and put the throw after the switch. In this way, if a case is not specified explicitly in the switch then the compiler will generate errors. Otherwise, the default case would handle it and we would know it at runtime only.

Reply:
Wouldn't that only help us if we did something like: hipSpMMAlgToHCCSpMMAlg(CUSPARSE_SPMM_CSR_ALG1)?
In reality we only ever call this like hipSpMMAlgToHCCSpMMAlg(alg) where alg is not known at runtime.

Reply:
It's an enum. When you switch an enum, the default case override any missing case.
Removing the default case will generate an error when a new field is added to the enum but is not handled explicitly by the switch. This way, if you touch the enum you will know at compile time everywhere you haven't specified the new case(s). Moreover, might be useful to use the constexpr for this routine since we could expect values known at compile time. But here if I add a new field to the enum, it will compile but tests will fail with the "Non existant .." exception. Compile time vs run time.

Reply:
Ok thanks. It looks like we do this a lot. Ill create a task to remove all these default cases in hipsparse.

Reply:
That is really smart of the compiler to omit an error if I add to the enum but dont have a default case. Like if I read alg from a file at runtime and was calling hipSpMMAlgToCudaSpMMAlg(alg) then the compiler cant know if my case is covered or not. So if I am understanding this correctly, a switch statement that that switches on a enum must cover all cases of the enum or a compile time error is generated?

Reply:
Ok very interesting. Just did a test and you are correct. I had no idea that is how it c++ works. It is able to detect that you may hit the throw and so gives a compile time error. Very cool

Reply:
No exactly, it will compile ... except if you activate -Wswitch, which throws warning on switch statements if all cases are not covered and the default statement is missing. -Wswitch is activated with -Wall and warnings can be transformed into errors with -Werror. 
So with -Wall -Werror, and if you remove the default statement, you'll have the compiler telling you if your switch is clean or not.
Possible valid cases are covered through the enumeration of all the cases (which is what we do all the time), the compiler helps with checking if you did, and the exception should be thrown outside the switch.